msdk-featdet/msdk-featdet-ADAP-3D/src/main/java/io/github/msdk/featdet/ADAP3D/common/algorithms/ADAP3DAlgorithm.java

```diff
@@ -0,0 +1,149 @@
+          - objSliceSparseMatrix.getRetentionTime(goodPeakList.get(i).lowerScanBound) / 60;
+      avgPeakWidth += peakWidth[i];
+      avgCoefOverArea += goodPeakList.get(i).objResult.coefOverArea;
+    }
+
+    avgPeakWidth = avgPeakWidth / goodPeakList.size();
+    avgCoefOverArea = avgCoefOverArea / goodPeakList.size();
+
+    int highestWaveletScale = (int) (avgPeakWidth * 60 / 2);
+    double coefOverAreaThreshold = avgCoefOverArea / 1.5;
+
+
+    List<Double> peakWidthList = Arrays.asList(ArrayUtils.toObject(peakWidth));
+    double peakDurationLowerBound = avgPeakWidth - LOW_BOUND_PEAK_WIDTH_PERCENT * avgPeakWidth;
+
+    double peakDurationUpperBound =
+        Collections.max(peakWidthList) + LOW_BOUND_PEAK_WIDTH_PERCENT * avgPeakWidth;
+
+    objParameters.setLargeScaleIn(highestWaveletScale);
+    objParameters.setMinPeakWidth(peakDurationLowerBound);
+    objParameters.setMaxPeakWidth(peakDurationUpperBound);
+    objParameters.setCoefAreaRatioTolerance(coefOverAreaThreshold);
+
+
+    List<PeakDetection.GoodPeakInfo> newGoodPeakList =
+        objPeakDetection.execute(objParameters, roundedFWHM);
+    List<SimpleFeature> newFeatureList = new ArrayList<SimpleFeature>();
+    newFeatureList.addAll(featureList);
+    newFeatureList.addAll(getFeature(newGoodPeakList));
+
+    return newFeatureList;
+  }
+
+
+
+  /**
+   * <p>
+   * This method takes list of GoodPeakInfo and returns list of type SimpleFeature. This method also
+   * builds Chromatogram for each good peak.
+   * </p>
+   * 
+   * @return featureList a list of {@link io.github.msdk.datamodel.impl.SimpleFeature}
+   */
+  private List<SimpleFeature> getFeature(List<PeakDetection.GoodPeakInfo> goodPeakList) {
+
+    List<SimpleFeature> featureList = new ArrayList<SimpleFeature>();
+    int lowerScanBound;
+    int upperScanBound;
+
+    for (int i = 0; i < goodPeakList.size(); i++) {
+
+      lowerScanBound = goodPeakList.get(i).lowerScanBound;
+      upperScanBound = goodPeakList.get(i).upperScanBound;
+      double mz = goodPeakList.get(i).mz;
+      float[] rtArray = objSliceSparseMatrix.getRetentionTimeArray(lowerScanBound, upperScanBound);
+      float[] intensityArray = objSliceSparseMatrix.getIntensities(goodPeakList.get(i));
+      double[] mzArray = new double[upperScanBound - lowerScanBound + 1];
+
+      for (int j = 0; j < upperScanBound - lowerScanBound + 1; j++) {
+        mzArray[j] = mz;
+      }
+
+      SimpleChromatogram chromatogram = new SimpleChromatogram();
+      chromatogram.setDataPoints(rtArray, mzArray, intensityArray,
+          upperScanBound - lowerScanBound + 1);
+
+      SimpleFeature feature = new SimpleFeature();
+      feature.setArea(CurveTool.normalize(intensityArray));
+      feature.setHeight(goodPeakList.get(i).maxHeight);
+      feature.setRetentionTime(
+          (float) objSliceSparseMatrix.getRetentionTime(goodPeakList.get(i).maxHeightScanNumber));
```

Comment:
This is an awful way of coding - first of all, it is impossible to understand why it is written in the way it is. Second thing is that if an exception is thrown on this line, you won't know if it was caused by goodPeakList.get(i), or by objSliceSparseMatrix.getRetentionTime(), or by conversion to float, etc.
Please avoid making many function calls in the same line of code. Keep the code simple and understandable.